### ISSUE TITLE
Alt-speed does not activate on schedule

### ISSUE THREAD

**Issue by phenomenomenom** [haugene/docker-transmission-openvpn#1128]
This issue has been created as requested and follows on from #981.

Many thanks to @haugene for adding tzdata to the image. This enables users to set the correct time on devices which can't map `/etc/timezone`, such as Synology. With the timezone now set correctly, transmission alt-speed should be usable. My problem is that alt-speed does not engage automatically.

I understand that timezone should be passed to the container with `-e TZ=Area/Location`. I also understand that the alt-speed times in the config should be 'minutes from midnight.'

The host is a Synology 918+ with user permissions set correctly. I am currently using tag `:latest` but I tested `:dev-alpine` and had the same results. For my setup, I am using docker-compose with an env file as I've ended up with a lot of variables to pass.

<details>
<summary>docker-compose:</summary>

```
version: '2.2'
services:
 transmission:
  image: haugene/transmission-openvpn
  container_name: transmission-openvpn
  cap_add:
    - NET_ADMIN
  devices:
    - /dev/net/tun
  restart: always
  networks:
    - transvpn_local
  ports:
    - "9100:9100"
    - "9101:9101"
  dns:
    - 8.8.8.8
    - 8.8.4.4
  volumes: # For other NAS users: Synology requires absolute volume mapping. Relative mapping will not work.
    - /volume1/docker/transmission-openvpn/resolv.conf:/etc/resolv.conf
    - /volume1/docker/transmission-openvpn/openvpn-custom/:/etc/openvpn/custom/
    - /volume1/Media/Transmission/:/data
  env_file:
    - /volume1/docker/transmission-openvpn/config.env
  environment:
    - OPENVPN_CONFIG=Ivacy_Custom_UDP

networks:
  transvpn_local: # I think the following network settings are unnecessary. I can't remember why I set it up this way.
    driver: bridge
    ipam:
      config:
        - subnet: 10.0.9.0/24
          gateway: 10.0.9.1
```
</details>

<details>
<summary>env file:</summary>

```
OPENVPN_PROVIDER=CUSTOM
OPENVPN_USERNAME=<removed>
OPENVPN_PASSWORD=<removed>
OPENVPN_OPTS=--inactive 3600 --ping 10 --ping-exit 60
LOCAL_NETWORK=10.0.0.0/23,10.0.5.0/24
CREATE_TUN_DEVICE=true
TZ=Australia/Sydney
PGID=<removed>
PUID=<removed>
UFW_DISABLE_IPTABLES_REJECT=true
TRANSMISSION_ALT_SPEED_DOWN=750
#TRANSMISSION_ALT_SPEED_ENABLED=false # See comment below.
TRANSMISSION_ALT_SPEED_TIME_BEGIN=1050
#TRANSMISSION_ALT_SPEED_TIME_DAY=127
TRANSMISSION_ALT_SPEED_TIME_ENABLED=true
TRANSMISSION_ALT_SPEED_TIME_END=1380
TRANSMISSION_ALT_SPEED_UP=400
TRANSMISSION_DOWNLOAD_QUEUE_SIZE=8
TRANSMISSION_PEER_LIMIT_GLOBAL=320
TRANSMISSION_PEER_LIMIT_PER_TORRENT=60
TRANSMISSION_RPC_AUTHENTICATION_REQUIRED=true
TRANSMISSION_RPC_ENABLED=true
TRANSMISSION_RPC_PASSWORD=<removed>
TRANSMISSION_RPC_PORT=9100
TRANSMISSION_RPC_USERNAME=<removed>
TRANSMISSION_RPC_WHITELIST=127.0.0.1,10.0.*.*
TRANSMISSION_RPC_WHITELIST_ENABLED=true
TRANSMISSION_SCRAPE_PAUSED_TORRENTS_ENABLED=false
TRANSMISSION_UTP_ENABLED=false
TRANSMISSION_WEB_UI=transmission-web-control
WEBPROXY_ENABLED=true
WEBPROXY_PORT=9101
DROP_DEFAULT_ROUTE=true
```
</details>
The above env is a copy of: https://github.com/haugene/docker-transmission-openvpn/blob/master/DockerEnv

I have removed unaltered lines for brevity. A few variables, such as TZ have been added compared with the original file.

One important thing to note is that setting `TRANSMISSION_ALT_SPEED_ENABLED=true` does not have any effect. Transmission neither starts in alt-speed nor does enabling this fix the scheduling problem.

After launching the container the web-ui indicates that alt-speed has been configured correctly:
![image](https://user-images.githubusercontent.com/59358618/77876958-bcb79f00-7297-11ea-9959-e591d613514c.png)

<details>
<summary>Here are my transmission settings.json</summary>

```
{
    "alt-speed-down": 750,
    "alt-speed-enabled": false,
    "alt-speed-time-begin": 1050,
    "alt-speed-time-day": 127,
    "alt-speed-time-enabled": true,
    "alt-speed-time-end": 1380,
    "alt-speed-up": 400,
    "bind-address-ipv4": "<removed>",
    "bind-address-ipv6": "::",
    "blocklist-enabled": false,
    "blocklist-url": "http://www.example.com/blocklist",
    "cache-size-mb": 4,
    "dht-enabled": true,
    "download-dir": "/data/completed",
    "download-limit": 100,
    "download-limit-enabled": 0,
    "download-queue-enabled": true,
    "download-queue-size": 8,
    "encryption": 1,
    "idle-seeding-limit": 30,
    "idle-seeding-limit-enabled": false,
    "incomplete-dir": "/data/incomplete",
    "incomplete-dir-enabled": true,
    "lpd-enabled": false,
    "max-peers-global": 200,
    "message-level": 2,
    "peer-congestion-algorithm": "",
    "peer-id-ttl-hours": 6,
    "peer-limit-global": 320,
    "peer-limit-per-torrent": 60,
    "peer-port": 51413,
    "peer-port-random-high": 65535,
    "peer-port-random-low": 49152,
    "peer-port-random-on-start": false,
    "peer-socket-tos": "default",
    "pex-enabled": true,
    "port-forwarding-enabled": false,
    "preallocation": 1,
    "prefetch-enabled": true,
    "queue-stalled-enabled": true,
    "queue-stalled-minutes": 30,
    "ratio-limit": 2,
    "ratio-limit-enabled": false,
    "rename-partial-files": true,
    "rpc-authentication-required": true,
    "rpc-bind-address": "0.0.0.0",
    "rpc-enabled": true,
    "rpc-host-whitelist": "",
    "rpc-host-whitelist-enabled": false,
    "rpc-password": "<removed>",
    "rpc-port": 9100,
    "rpc-url": "/transmission/",
    "rpc-username": "<removed>",
    "rpc-whitelist": "127.0.0.1,10.0.*.*",
    "rpc-whitelist-enabled": true,
    "scrape-paused-torrents-enabled": false,
    "script-torrent-done-enabled": false,
    "script-torrent-done-filename": "",
    "seed-queue-enabled": false,
    "seed-queue-size": 10,
    "speed-limit-down": 100,
    "speed-limit-down-enabled": false,
    "speed-limit-up": 100,
    "speed-limit-up-enabled": false,
    "start-added-torrents": true,
    "trash-original-torrent-files": false,
    "umask": 2,
    "upload-slots-per-torrent": 14,
    "utp-enabled": false,
    "watch-dir": "/data/watch",
    "watch-dir-enabled": true,
    "watch-dir-force-generic": false
}
```
</details>

I tested further by SSH'ing into the container with `docker exec -it transmission-openvpn /bin/bash`.
On `:latest`, command `dpkg --list tzdata` returned that version 2019c-0ubuntu0.18.04 was installed.
On `:dev-alpine`, command `apk info tzdata` returned that tzdata-2019c-r0 was installed
On both versions, `date` returned my time, and zone 'AEDT' which is correct.

**Comment by phenomenomenom** [comment]
A few more details to add in response to [haugene's post](https://github.com/haugene/docker-transmission-openvpn/issues/981#issuecomment-610247191):

The settings.json in my previous post was captured while the container was running, before changing any settings.

If I enable alt-speed through the web-ui, then stop the container and view settings.json, there are no changes. When I reboot the container, alt-speed is disabled which you'd expect.

Changing other variables via the web-ui had the same result. They're not writing to settings.json.

As suggested, I exec'd into the container to try to stop transmission but this is where I got a bit stuck.
Command: `/etc/init.d/transmission-daemon stop`
Result: ` * Stopping bittorrent daemon transmission-daemon`

Unfortunately, either transmission isn't actually being stopped or it is restarting immediately. I could well be doing something wrong as this is a bit beyond my ability. Do you have suggestions for what I could try?

In any case, settings.json is not changing at any point while running, after stopping the container, or after attempting to stop the service.

**Comment by haugene** [comment]
Thanks for a well documented issue. 

To stop Transmission and make it persist the settings I exec into the container and run `ps aux`. Then I find the PID of the transmission-daemon, and run `kill <PID>` that will signal it to stop.

```
root@4db4a68d0b8a:/# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0    212     4 ?        Ss   08:10   0:00 dumb-init /etc/openvpn/start.sh
root         7  0.0  0.3  42340  7008 ?        Ss   08:10   0:00 openvpn --script-security 2 --up-delay --up /etc/openvpn/tunnelUp.sh --down /etc/openvpn/tunnelDown.sh --config /e
root        43  0.0  0.3 350220  7420 ?        Ssl  08:10   0:00 /usr/bin/transmission-daemon -g /data/transmission-home --logfile /data/transmission-home/transmission.log
root        80  3.0  0.1  18508  3368 pts/0    Ss   08:25   0:00 bash
root        95  0.0  0.1  34408  2864 pts/0    R+   08:25   0:00 ps aux
root@4db4a68d0b8a:/# kill 43
```

I did this now and see that when I enable alt-speed (clicking the turtle) from Transmission and then stop it I see the change as expected:
```
"alt-speed-enabled": false,

... changed to ...

"alt-speed-enabled": true,
```

In addition. Changing that property in settings.json and starting Transmission from within the container I see that Transmission behaves correctly. So it's not them, it's us. 😄 
At leas it seems that way for now. And we need to investigate a bit further...

**Comment by haugene** [comment]
When I set the `TRANSMISSION_ALT_SPEED_ENABLED=true` it actually works here 🤔 Well, it shows as activated in Transmission UI at least.

If I see correctly you're not using the standard Transmission UI? I see the turtle mode being activated and deactivated based on my value in alt-speed-enabled passed through environment variables. Can you try the standard UI just in case?

If not. Try this docker-compose. You should be able to run it in parallel with the existing one as I haven't mounted the data dir and is using 8080 port on the proxy.

```
version: '2.2'
services:
  transmission:
    image: haugene/transmission-openvpn
    container_name: transmission # linked to this name from proxy container
    cap_add:
      - NET_ADMIN
    restart: on-failure
    dns:
      - 8.8.8.8
      - 8.8.4.4
    volumes:
      - /volume1/docker/transmission-openvpn/resolv.conf:/etc/resolv.conf:ro
      - /volume1/docker/transmission-openvpn/openvpn-custom/:/etc/openvpn/custom/:ro
    environment:
      - OPENVPN_CONFIG=Ivacy_Custom_UDP
      - OPENVPN_PROVIDER=CUSTOM
      - OPENVPN_USERNAME=<removed>
      - OPENVPN_PASSWORD=<removed>
      - OPENVPN_OPTS=--inactive 3600 --ping 10 --ping-exit 60
      - CREATE_TUN_DEVICE=true

  # Instead of LOCAL_NETWORK
  transmission-web-proxy:
    image: haugene/transmission-openvpn-proxy
    ports:
      - 8080:8080

```

A couple of comments to your docker-compose that I saw now that I used it.
First I like to mount stuff as read-only if the container shouldn't be able to write to it. So I did that on resolv.conf and openvpn config.
Second, you should either set CREATE_TUN_DEVICE=true or mount /dev/net/tun. You previously had to mount it, but now it can be created from inside the image.

**Comment by pkishino** [comment]
Hello, I tested the new TZ, which works fine more :) also scheduling and alt-speed pools like it’s working for me (OS X docker)

**Comment by edgd1er** [comment]
Well, I thought it was ok, but not in fact....

login in the container, the date is properly define, env $TZ also
```
root@transmission:/# date
Thu May  7 18:55:40 CEST 2020

root@transmission:/# echo $TZ
Europe/Paris
```
but the alt-speed is still on, one hour after it should have stopped.

if I reconfigure tzdata, the time is wrongly set and time zone is set to UTC.
```
root@transmission:/# dpkg-reconfigure tzdata
Current default time zone: 'Etc/UTC'
Local time is now:      Thu May  7 16:57:13 UTC 2020.
Universal Time is now:  Thu May  7 16:57:13 UTC 2020.
```

linking the proper timezone to localetime and reconfiguring, fix the time shift:
```
root@transmission:/# ln -sf /usr/share/zoneinfo/Europe/Paris /etc/localtime
root@transmission:/# dpkg-reconfigure tzdata

Current default time zone: 'Europe/Paris'
Local time is now:      Thu May  7 18:58:05 CEST 2020.
Universal Time is now:  Thu May  7 16:58:05 UTC 2020.
```

I guess the link is needed eventually.

**Comment by pkishino** [comment]
well, if there is any further update on this we can re-open

**Comment by superkrups20056** [comment]
I can confirm that altspeed does not respect the timezone environmental variable and still unfortunately goes off of UTC. I set my time zone to TZ=America/New_York. Using the transmission-web-gui, I set the altspeed cutoff time to a minute after my time in UTC and it promptly changes from a turtle to a rocket on refresh of the webpage when the minute passes. I'm running the latest stable build as of today on a Synology NAS.
